---
title: "Differential abundance of B-cell subtypes" 
output:
  html_document:
    df_print: paged
header-includes:
- \usepackage[colorinlistoftodos]{todonotes}
- \hypersetup{colorlinks=true,linkcolor=blue!30!black}
- \geometry{rmargin=1.5in}
---

\textbf{The first part of this is your code copy and pasted}. 

## Test two methods of differential expression 
1. Tree based
2. Current annotation vs. every other annotation 

## Tree Based
1. Hierarchical clustering of psuedo-bulk at fine and main level
2. For each node in the tree perform differential expression analysis using limma
3. Check if read count from RNA-seq are available and deg using deseq2

```{r, warning = FALSE}
library(SingleCellExperiment)
library(dplyr)
library(data.table)
library(SummarizedExperiment, BiocManager)
library(Seurat)
library(magrittr)
# singler.dataset = readRDS("~/liulab_home/data/single_cell/singleR.dataset.RDS")
# ref.data = singler.dataset$MonacoImmuneData
library(SingleR)
ref.data <-  SingleR::MonacoImmuneData()
col.dat = ref.data@colData %>% as.data.table %>% 
  .[,label.fine:=as.factor(label.fine)]  %>% 
  .[,label.main:=as.factor(label.main)]
create.psuedobulk <- function(exp.mat, labels) {
  dt1 = data.table(V1=labels)
  one.hot = mltools::one_hot(dt1) %>% as.matrix
  one.hot %>% colMeans %>% 
    sweep(one.hot,., MARGIN = 2, FUN = "/")
  out = exp.mat %*% one.hot
  out %>% set_colnames(gsub("V1_", colnames(one.hot), replacement = ""))
}
ref.psuedo.main = create.psuedobulk(ref.data@assays@data$logcounts, col.dat$label.main)
d = 1 -cor(ref.psuedo.main[,1:3], method = "spearman")
hc1 <- hclust(dist(d), method = "complete" )
plot(hc1, cex = 0.6, hang = -1, main="Monaco main")
ref.psuedo.fine = create.psuedobulk(ref.data@assays@data$logcounts, col.dat$label.fine)
d = 1 -cor(ref.psuedo.fine, method = "spearman")
hc1 <- hclust(dist(d), method = "complete" )
plot(hc1, cex = 0.6, hang = -1, main="Monaco fine")
```
2. For each node in the tree perform differential expression analysis using limma
```{r, echo = TRUE, results = 'hide'}
library(limma)
find.variable.genes = function(exp.curr,
                                   selection.method = "vst",
  loess.span = 0.3,
  clip.max = 'auto',
  mean.function = FastExpMean,
  dispersion.function = FastLogVMR,
  num.bin = 20,
  binning.method = "equal_width",
  nfeatures = 2000,
   mean.cutoff = c(0.1, 8),
  dispersion.cutoff = c(1, Inf),
  verbose = TRUE,
  ...
  ) {
   hvf.info <- Seurat::FindVariableFeatures(
    object = exp.curr,
    selection.method = selection.method,
    loess.span = loess.span,
    clip.max = clip.max,
    mean.function = mean.function,
    dispersion.function = dispersion.function,
    num.bin = num.bin,
    binning.method = binning.method,
    verbose = verbose
  )
   
  hvf.info <- hvf.info[which(x = hvf.info[, 1, drop = TRUE] != 0), ]
  if (selection.method == "vst") {
    hvf.info <- hvf.info[order(hvf.info$vst.variance.standardized, decreasing = TRUE), , drop = FALSE]
  } else {
    hvf.info <- hvf.info[order(hvf.info$mvp.dispersion, decreasing = TRUE), , drop = FALSE]
  }
  selection.method <- switch(
    EXPR = selection.method,
    'mvp' = 'mean.var.plot',
    'disp' = 'dispersion',
    selection.method
  )
  top.features <- switch(
    EXPR = selection.method,
    'mean.var.plot' = {
      means.use <- (hvf.info[, 1] > mean.cutoff[1]) & (hvf.info[, 1] < mean.cutoff[2])
      dispersions.use <- (hvf.info[, 3] > dispersion.cutoff[1]) & (hvf.info[, 3] < dispersion.cutoff[2])
      rownames(x = hvf.info)[which(x = means.use & dispersions.use)]
    },
    'dispersion' = head(x = rownames(x = hvf.info), n = nfeatures),
    'vst' = head(x = rownames(x = hvf.info), n = nfeatures),
    stop("Unkown selection method: ", selection.method)
  )
  top.features
}


mydeg <- function(exp.curr, groups) {
  genes.sel = find.variable.genes(exp.curr) 
  exp.curr.sub = exp.curr[rownames(exp.curr) %in% genes.sel,]
  # dge <- DGEList(counts=expression.mat)
  # logCPM <- cpm(dge, log=TRUE, prior.count=3)
  logCPM = log1p(exp.curr.sub)
  fit <- lmFit(logCPM, design = groups)
  fit <- eBayes(fit, trend=TRUE)
  deg =  topTable(fit, coef=ncol(design), 2000)
  # browser()
  out = deg %>% data.table %>% 
    .[,gene:=rownames(deg)]
  out
}
 
    

#' Get differential expression in reference dataset tree. 
#'
#' @param exp.all Expression matrix of all sample (genes x samples)
#' @param tree tree of hclust object
#' @param labels Annotation of samples (columns) of exp.all 
#'
#' @return list with two elements 1. Differential expression in the tree internal nodes 2. leaf.child : leaves children of each internal node  
#' @export
#'
#' @examples
create.ref.tree.deg  <- function(exp.all, tree, labels) {
  # tree$merge
  stopifnot(length(labels)==ncol(exp.all))
  degs = leaf.child = list()
  # exp.all = assay(object)
  cntr <- 1
  for (parent in seq(nrow(tree$merge))) {
    #if(cntr == 1) {cntr <- cntr + 1; next}
    parent.chr = as.character(parent)
    childs = tree$merge[parent,]
    groups = rep(NA, ncol(exp.all))
    for (child.inx in seq_along(childs)) {
      print(child.inx)
      child = childs[child.inx]
      child.chr = as.character(child)
      if(child < 0) {
        child = -1 * child
        leaf.child[[child.chr]] = tree$labels[child]
      }
      leaf.child[[parent.chr]] = unlist(c( leaf.child[[parent.chr]], leaf.child[[child.chr]]))
      groups[(labels %in% leaf.child[[parent.chr]]) & (is.na(groups))] = child.inx-1
      print(groups)
    }
    na.inx = which(!is.na(groups))
    print(na.inx)
    exp.curr = exp.all[,na.inx]
    groups.curr = groups[na.inx]
    print(groups.curr)
    degs[[parent.chr]] = mydeg(exp.curr, groups.curr)
  }
  list(degs=degs, leaf.child=leaf.child)
}
# aa = Seurat::FindVariableFeatures(assay(ref.data$ref.se))
# aa = find.varialable.genes(assay(ref.data$ref.se)[1:10000,1:10])
```


```{r}
find.variable.genes = function(exp.curr,
                                   selection.method = "vst",
  loess.span = 0.3,
  clip.max = 'auto',
  mean.function = FastExpMean,
  dispersion.function = FastLogVMR,
  num.bin = 20,
  binning.method = "equal_width",
  nfeatures = 2000,
   mean.cutoff = c(0.1, 8),
  dispersion.cutoff = c(1, Inf),
  verbose = TRUE,
  ...
  ) {
   hvf.info <- Seurat::FindVariableFeatures(
    object = exp.curr,
    selection.method = selection.method,
    loess.span = loess.span,
    clip.max = clip.max,
    mean.function = mean.function,
    dispersion.function = dispersion.function,
    num.bin = num.bin,
    binning.method = binning.method,
    verbose = verbose
  )
   
  hvf.info <- hvf.info[which(x = hvf.info[, 1, drop = TRUE] != 0), ]
  if (selection.method == "vst") {
    hvf.info <- hvf.info[order(hvf.info$vst.variance.standardized, decreasing = TRUE), , drop = FALSE]
  } else {
    hvf.info <- hvf.info[order(hvf.info$mvp.dispersion, decreasing = TRUE), , drop = FALSE]
  }
  selection.method <- switch(
    EXPR = selection.method,
    'mvp' = 'mean.var.plot',
    'disp' = 'dispersion',
    selection.method
  )
  top.features <- switch(
    EXPR = selection.method,
    'mean.var.plot' = {
      means.use <- (hvf.info[, 1] > mean.cutoff[1]) & (hvf.info[, 1] < mean.cutoff[2])
      dispersions.use <- (hvf.info[, 3] > dispersion.cutoff[1]) & (hvf.info[, 3] < dispersion.cutoff[2])
      rownames(x = hvf.info)[which(x = means.use & dispersions.use)]
    },
    'dispersion' = head(x = rownames(x = hvf.info), n = nfeatures),
    'vst' = head(x = rownames(x = hvf.info), n = nfeatures),
    stop("Unkown selection method: ", selection.method)
  )
  top.features
}


mydeg <- function(exp.curr, groups, use.log=FALSE, num.var.genes =2000) {
  require(Matrix)
  require(dplyr)
  require(limma)
  
  groups[groups == 0] <- -1
 
  genes.sel = find.variable.genes(exp.curr, nfeatures = num.var.genes) 
  exp.curr.sub = exp.curr[rownames(exp.curr)%in%  genes.sel,]
  # dge <- DGEList(counts=expression.mat)
  # logCPM <- cpm(dge, log=TRUE, prior.count=3)
  if(use.log) exp.curr.sub = log1p(exp.curr.sub)
  design <- model.matrix(~ 0+factor(groups))
  colnames(design) <- c("group1", "group2")
  fit <- lmFit(exp.curr.sub, design)
  contrast.matrix <- makeContrasts(group2-group1, levels=design)
  fit2 <- contrasts.fit(fit, contrast.matrix)
  fit <- eBayes(fit)
  deg =  topTable(fit, coef=1, 2000)
  # browser()
  out = deg %>% data.table %>% 
    .[,gene:=rownames(deg)]
  out
}
 
    

#' Get differential expression in reference dataset tree. 
#'
#' @param exp.all Expression matrix of all sample (genes x samples)
#' @param tree tree of hclust object
#' @param labels Annotation of samples (columns) of exp.all 
#'
#' @return list with two elements 1. Differential expression in the tree internal nodes 2. leaf.child : leaves children of each internal node  
#' @export
#'
#' @examples
create.ref.tree.deg  <- function(exp.all, tree, labels, ncores = 4) {
  # tree$merge
  stopifnot(length(labels)==ncol(exp.all))
  degs = leaf.child = list()
  groups.ls = list()
  # exp.all = assay(object)
  for (parent in seq(nrow(tree$merge))) {
    parent.chr = as.character(parent)
    childs = tree$merge[parent,]
    groups = rep(NA, ncol(exp.all))
    for (child.inx in seq_along(childs)) {
      child = childs[child.inx]
      child.chr = as.character(child)
      if(child < 0){
        child = -1 * child
        leaf.child[[child.chr]] = tree$labels[child]
      }
      leaf.child[[parent.chr]] = unlist(c( leaf.child[[parent.chr]], leaf.child[[child.chr]]))
      groups[labels %in% leaf.child[[child.chr]]] = child.inx-1
    }
    groups.ls[[parent.chr]] = groups
  }

  out <- list() 
  out1 <- list()
  for(parent in seq(nrow(tree$merge))) {
    print(parent)
    parent.chr = as.character(parent)
    groups = groups.ls[[parent.chr]]
    na.inx = which(!is.na(groups))
    exp.curr = exp.all[,na.inx]
    groups.curr = groups[na.inx]
    print(groups.curr)
    out[[parent]] <- mydeg(exp.curr, groups.curr)
    groups.curr <- 1 - groups.curr
    print(groups.curr)
    out1[[parent]] <-  mydeg(exp.curr, groups.curr)
  }

 names(out) = names(groups.ls)
 names(out1) = names(groups.ls)
  return(list(degsR=out, degL = out1, leaf.child=leaf.child))
}
```

```{r}
find.variable.genes = function(exp.curr,
                                   selection.method = "vst",
  loess.span = 0.3,
  clip.max = 'auto',
  mean.function = FastExpMean,
  dispersion.function = FastLogVMR,
  num.bin = 20,
  binning.method = "equal_width",
  nfeatures = 2000,
   mean.cutoff = c(0.1, 8),
  dispersion.cutoff = c(1, Inf),
  verbose = TRUE,
  ...
  ) {
   hvf.info <- Seurat::FindVariableFeatures(
    object = exp.curr,
    selection.method = selection.method,
    loess.span = loess.span,
    clip.max = clip.max,
    mean.function = mean.function,
    dispersion.function = dispersion.function,
    num.bin = num.bin,
    binning.method = binning.method,
    verbose = verbose
  )
   
  hvf.info <- hvf.info[which(x = hvf.info[, 1, drop = TRUE] != 0), ]
  if (selection.method == "vst") {
    hvf.info <- hvf.info[order(hvf.info$vst.variance.standardized, decreasing = TRUE), , drop = FALSE]
  } else {
    hvf.info <- hvf.info[order(hvf.info$mvp.dispersion, decreasing = TRUE), , drop = FALSE]
  }
  selection.method <- switch(
    EXPR = selection.method,
    'mvp' = 'mean.var.plot',
    'disp' = 'dispersion',
    selection.method
  )
  top.features <- switch(
    EXPR = selection.method,
    'mean.var.plot' = {
      means.use <- (hvf.info[, 1] > mean.cutoff[1]) & (hvf.info[, 1] < mean.cutoff[2])
      dispersions.use <- (hvf.info[, 3] > dispersion.cutoff[1]) & (hvf.info[, 3] < dispersion.cutoff[2])
      rownames(x = hvf.info)[which(x = means.use & dispersions.use)]
    },
    'dispersion' = head(x = rownames(x = hvf.info), n = nfeatures),
    'vst' = head(x = rownames(x = hvf.info), n = nfeatures),
    stop("Unkown selection method: ", selection.method)
  )
  top.features
}


mydeg <- function(exp.curr, groups, use.log=FALSE, num.var.genes =2000) {
  require(Matrix)
  require(dplyr)
  require(limma)
 
  genes.sel = find.variable.genes(exp.curr, nfeatures = num.var.genes) 
  exp.curr.sub = exp.curr[rownames(exp.curr)%in%  genes.sel,]
  # dge <- DGEList(counts=expression.mat)
  # logCPM <- cpm(dge, log=TRUE, prior.count=3)
  if(use.log) logCPM = log1p(exp.curr.sub)
  design <- model.matrix(~ 0+factor(groups))
  colnames(design) <- c("group1", "group2")
  contrast.matrix <- makeContrasts(group2-group1, levels=design)
  fit <- lmFit(exp.curr.sub, design = design)
  fit2 <- contrasts.fit(fit, contrast.matrix)
  fit <- eBayes(fit2)
  deg =  topTable(fit, coef=1, 2000)
  # browser()
  out = deg %>% data.table %>% 
    .[,gene:=rownames(deg)]
  out
}
 
    

#' Get differential expression in reference dataset tree. 
#'
#' @param exp.all Expression matrix of all sample (genes x samples)
#' @param tree tree of hclust object
#' @param labels Annotation of samples (columns) of exp.all 
#'
#' @return list with two elements 1. Differential expression in the tree internal nodes 2. leaf.child : leaves children of each internal node  
#' @export
#'
#' @examples
create.ref.tree.deg  <- function(exp.all, tree, labels, ncores = 32) {
  # tree$merge
  stopifnot(length(labels)==ncol(exp.all))
  degs = leaf.child = list()
  groups.ls = list()
  # exp.all = assay(object)
  for (parent in seq(nrow(tree$merge))) {
    parent.chr = as.character(parent)
    childs = tree$merge[parent,]
    groups = rep(NA, ncol(exp.all))
    for (child.inx in seq_along(childs)) {
      child = childs[child.inx]
      child.chr = as.character(child)
      if(child < 0){
        child = -1 * child
        leaf.child[[child.chr]] = tree$labels[child]
      }
      leaf.child[[parent.chr]] = unlist(c( leaf.child[[parent.chr]], leaf.child[[child.chr]]))
      groups[labels %in% leaf.child[[child.chr]]] = child.inx-1
    }
    groups.ls[[parent.chr]] = groups
  }
require(doMC)
require(foreach)
registerDoMC(cores = ncores)
degs = foreach( parent = seq(nrow(tree$merge)), .inorder=T) %dopar%{
    parent.chr = as.character(parent)
    groups = groups.ls[[parent.chr]]
    na.inx = which(!is.na(groups))
    exp.curr = exp.all[,na.inx]
    groups.curr = groups[na.inx]
    out = tryCatch({
      mydeg(exp.curr, groups.curr)
    }, 
    error = function(e) NA)
    return( out )
}
 names(degs) = names(groups.ls)
  list(degs=degs, leaf.child=leaf.child)
}
# aa = Seurat::FindVariableFeatures(assay(ref.data$ref.se))
# aa = find.varialable.genes(assay(ref.data$ref.se)[1:10000,1:10])
```





```{r, echo = TRUE, results = 'hide'}
out = create.ref.tree.deg(exp.all=ref.data@assays@data$logcounts, tree=hc1, labels=col.dat$label.fine)
```



## B cell subtypes

The Monaco dataset has $4$ B-cell subpopulations: Naive, Non-switched memory, exhausted, switched memory. Let's load the data 

```{r}
Sco <- readRDS("/Users/phillipnicol/Desktop/local_files/data/scDiffPopDATA/gse145281_tumor/gse145281T_manualAnnot.RDS")
```


Let's subset to B-cells only 


```{r}
Sco <- Sco[,Sco@meta.data$CellType == "B"]
```

Do a differential expression between blood and tumor

```{r}
Idents(Sco) <- Sco@meta.data$Type
markers <- FindMarkers(Sco, ident.1 = "Blood")
markers <- markers[markers$p_val_adj < 0.05,]
BloodMarkers <- rownames(markers[markers$avg_logFC > 0,])
TumorMarkers <- rownames(markers[markers$avg_logFC < 0, ])
```


Make Contingency table for each subtype 

## Children of B cells (Internal nodes with 2 leaves)
```{r}
#NAIVE B AND NON-SWICHED MEMORY 
deg_naive_b <- out$degL$`14`[out$degL$`14`$logFC > 0]
deg_naive_b <- out$degL$`14`[out$degL$`14`$adj.P.Val < 0.05]$gene

a <- length(which(BloodMarkers %in% deg_naive_b))
b <- length(which(TumorMarkers %in% deg_naive_b))
c <- length(BloodMarkers) - a
d <- length(TumorMarkers) - b
C <- matrix(c(a,b,c,d), nrow = 2, ncol = 2, byrow = TRUE)
rownames(C) <- c("Markers of subtype", "Not markers of subtype")
colnames(C) <- c("Upregulated in blood", "Upregulated in tumor")
print(C)
paste("Enriched in blood, p-val:", fisher.test(C, alternative = "greater")$p.val)
paste("Enriched in tumor, p-val:", fisher.test(C, alternative = "less")$p.val)



#EXHAUSTED B-CELL AND SWITCHED MEMORY B-CELL
deg_naive_b <- out$degsR$`14`[out$degsR$`14`$logFC > 0]
deg_naive_b <- out$degsR$`14`[out$degsR$`14`$adj.P.Val < 0.05]$gene

a <- length(which(BloodMarkers %in% deg_naive_b))
b <- length(which(TumorMarkers %in% deg_naive_b))
c <- length(BloodMarkers) - a
d <- length(TumorMarkers) - b
C <- matrix(c(a,b,c,d), nrow = 2, ncol = 2, byrow = TRUE)
rownames(C) <- c("Markers of subtype", "Not markers of subtype")
colnames(C) <- c("Upregulated in blood", "Upregulated in tumor")
print(C)
paste("Enriched in blood, p-val:", fisher.test(C, alternative = "greater")$p.val)
paste("Enriched in tumor, p-val:", fisher.test(C, alternative = "less")$p.val)

```



## Leaf nodes 
```{r}
#Naive B-cells
deg_naive_b <- out$degL$`5`[out$degL$`5`$logFC > 0]
deg_naive_b <- out$degL$`5`[out$degL$`5`$adj.P.Val < 0.05]$gene

a <- length(which(BloodMarkers %in% deg_naive_b))
b <- length(which(TumorMarkers %in% deg_naive_b))
c <- length(BloodMarkers) - a
d <- length(TumorMarkers) - b
C <- matrix(c(a,b,c,d), nrow = 2, ncol = 2, byrow = TRUE)
rownames(C) <- c("Markers of subtype", "Not markers of subtype")
colnames(C) <- c("Upregulated in blood", "Upregulated in tumor")
print(C)
paste("Enriched in blood, p-val:", fisher.test(C, alternative = "greater")$p.val)
paste("Enriched in tumor, p-val:", fisher.test(C, alternative = "less")$p.val)

#Non-switched memory B
deg_nsm_b <- out$degsR$`5`[out$degsR$`5`$logFC > 0]
deg_nsm_b <- out$degsR$`5`[out$degsR$`5`$adj.P.Val< 0.05]$gene

a <- length(which(BloodMarkers %in% deg_nsm_b))
b <- length(which(TumorMarkers %in% deg_nsm_b))
c <- length(BloodMarkers) - a
d <- length(TumorMarkers) - b
C <- matrix(c(a,b,c,d), nrow = 2, ncol = 2, byrow = TRUE)
rownames(C) <- c("Markers of subtype", "Not markers of subtype")
colnames(C) <- c("Upregulated in blood", "Upregulated in tumor")
print(C)
paste("Enriched in blood, p-val:", fisher.test(C, alternative = "greater")$p.val)
paste("Enriched in tumor, p-val:", fisher.test(C, alternative = "less")$p.val)

#Exhausted B
deg_e_b <- out$degL$`11`[out$degL$`11`$logFC > 0]
deg_e_b <- out$degL$`11`[out$degL$`11`$adj.P.Val < 0.05]$gene

a <- length(which(BloodMarkers %in% deg_e_b))
b <- length(which(TumorMarkers %in% deg_e_b))
c <- length(BloodMarkers) - a
d <- length(TumorMarkers) - b
C <- matrix(c(a,b,c,d), nrow = 2, ncol = 2, byrow = TRUE)
rownames(C) <- c("Markers of subtype", "Not markers of subtype")
colnames(C) <- c("Upregulated in blood", "Upregulated in tumor")
print(C)
paste("Enriched in blood, p-val:", fisher.test(C, alternative = "greater")$p.val)
paste("Enriched in tumor, p-val:", fisher.test(C, alternative = "less")$p.val)


#Switched mem B
deg_sm_b <- out$degsR$`11`[out$degsR$`11`$logFC < 0]
deg_sm_b <- out$degsR$`11`[out$degsR$`11`$adj.P.Val < 0.05]$gene

a <- length(which(BloodMarkers %in% deg_sm_b))
b <- length(which(TumorMarkers %in% deg_sm_b))
c <- length(BloodMarkers) - a
d <- length(TumorMarkers) - b
C <- matrix(c(a,b,c,d), nrow = 2, ncol = 2, byrow = TRUE)
rownames(C) <- c("Markers of subtype", "Not markers of subtype")
colnames(C) <- c("Upregulated in blood", "Upregulated in tumor")
print(C)
paste("Enriched in blood, p-val:", fisher.test(C, alternative = "greater")$p.val)
paste("Enriched in tumor, p-val:", fisher.test(C, alternative = "less")$p.val)


```


